cpp/src/binning/bin.cu

```diff
@@ -16,17 +16,179 @@
 #include <cudf/column/column_factories.hpp>
 #include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
 #include <cudf/utilities/error.hpp>
+#include <cudf/types.hpp>
 #include <cudf/utilities/type_dispatcher.hpp>
+#include <thrust/functional.h>
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <cudf/detail/valid_if.cuh>
+#include <thrust/pair.h>
 
 namespace cudf {
 
-namespace bin {
+// Sentinel used to indicate that an input value should be placed in the null
+// bin.
+// NOTE: In theory if a user decided to specify 2^31 bins this would fail. We
+// could make this an error in Python, but that is such a crazy edge case...
+constexpr size_type NULL_VALUE{-1};
 
-// Bin the input by the edges in left_edges and right_edges.
+namespace detail {
+namespace {
+
+template <typename T, typename StrictWeakOrderingLeft, typename StrictWeakOrderingRight>
+struct bin_finder
+{
+    bin_finder(
+            const T *left_edges,
+            const T *left_edges_end,
+            const T *right_edges
+            )
+        : m_left_edges(left_edges), m_left_edges_end(left_edges_end), m_right_edges(right_edges)
+    {}
+
+    __device__ size_type operator()(thrust::pair<T, bool> input_value) const
+    {
+        if (!input_value.second)
+            return NULL_VALUE;
+
+        const T value = input_value.first;
+        auto bound = thrust::lower_bound(thrust::seq,
+                m_left_edges, m_left_edges_end,
+                value,
+                m_left_comp);
+
+        // Exit early and return NULL for values not within the interval.
+        if ((bound == m_left_edges) || (bound == m_left_edges_end))
+            return NULL_VALUE;
+
+        // We must subtract 1 because lower bound returns the first index
+        // _greater than_ the value. This is safe because bound == m_left edges
+        // would already have triggered a NULL return above.
+        auto index = bound - m_left_edges - 1;
+        return (m_right_comp(value, m_right_edges[index])) ? index : NULL_VALUE;
+    }
+
+    const T *m_left_edges{};  // Pointer to the beginning of the device data containing left bin edges.
+    const T *m_left_edges_end{};  // Pointer to the end of the device data containing left bin edges.
+    const T *m_right_edges{};  // Pointer to the beginning of the device data containing right bin edges.
+    // TODO: Can I implement these as static members rather than making an instance on construction?
+    StrictWeakOrderingLeft m_left_comp{}; // Comparator used for left edges.
+    StrictWeakOrderingRight m_right_comp{}; // Comparator used for left edges.
+};
+
+
+struct filter_null_sentinel
+{
+    __device__ bool operator()(size_type i)
+    {
+        return (i != NULL_VALUE);
+    }
+};
+
+/// Bin the input by the edges in left_edges and right_edges.
+template <typename T, typename StrictWeakOrderingLeft, typename StrictWeakOrderingRight, bool InputIsNullable>
```

Comment:
The naming here implies the comparison operator is a strict, weak ordering operator, but less_equal/greater_equal aren't strict, weak ordering. Does it really need to be a strict, weak ordering op?

Reply:
When it doubt, I usually look at the docs for the equivalent STL algorithm. Looking at `[std::lower_bound`](https://en.cppreference.com/w/cpp/algorithm/lower_bound), it looks like the comparator need not be strict. In fact, it need not even satisfy the "Compare" concept. 

So I think you're safe in using those operators, I'd just rename them to avoid confusion (mainly mine).